### ISSUE TITLE
question about computing contribution utility

### ISSUE THREAD

**Issue by kiranchari** [shibhansh/loss-of-plasticity#9]
Hi, I am trying to understand the calculation of utility in `cbp_conv.py`: 

https://github.com/shibhansh/loss-of-plasticity/blob/63c35f3c758bbb713dd42c72d43dc192fde0d109/lop/algos/cbp_conv.py#L89
https://github.com/shibhansh/loss-of-plasticity/blob/63c35f3c758bbb713dd42c72d43dc192fde0d109/lop/algos/cbp_conv.py#L90

Based on the call to `mean(dim=(0,2,3))`, it appears that the utility is computed per channel in the conv2d output rather than per neuron - is this correct? However, equation 1 in the paper (https://www.nature.com/articles/s41586-024-07711-7#Sec6) mentions that the utility is computed per neuron in a layer.

Could you please elaborate on the utility computation and clarify if my understanding is correct? Also, could you comment on why the utility is computed per neuron rather than per channel?

Thanks

**Comment by kiranchari** [comment]
@shibhansh I have a question about the utility calculation. I see that the mean of the weights across dimensions (0,2,3) is computed as follows: 

https://github.com/shibhansh/loss-of-plasticity/blob/63c35f3c758bbb713dd42c72d43dc192fde0d109/lop/algos/cbp_conv.py#L89

The Conv2d weight tensor has shape `(in_channels, out_channels, kernel_size, kernel_size)`. So the above mean is computed over `in_channels, kernel_size, kernel_size` and should have a final shape of `(out_channels)`. 

The mean over features is computed as follows:

https://github.com/shibhansh/loss-of-plasticity/blob/63c35f3c758bbb713dd42c72d43dc192fde0d109/lop/algos/cbp_conv.py#L90

Assuming self.features has shape `(batch_size, in_channels, height, width)`, the above mean is averaging across `batch_size, height and width` and should have a final shape of `(in_channels)`.

It appears to me there is a mismatch in the dimensions that are being multiplied to compute utility i.e. `out_channels` vs `in_channels`. It doesn't raise a runtime error because these two dimensions happen to be the same (due to the pooling operation: https://github.com/shibhansh/loss-of-plasticity/blob/63c35f3c758bbb713dd42c72d43dc192fde0d109/lop/nets/conv_net2.py#L52)

Is my understanding correct? Please let me know if I am missing something.

Thanks!

**Comment by shibhansh** [comment]
Hi Kiran, 

Answer to your first question:
Yes, utility is computed per channel in the conv2d output. However, note that in a convolutional network, the equivalent of a neuron is a channel. So, there is no difference between a neuron's and a channel's utility. I hope this clarifies your confusion.

Answer to your second question:
Although, Conv2d takes inputs _(in_channels, out_channels, kernel_size ..)_,  the Conv2d weight tensor has shape _(out_channels, in_channels, kernel_size, kernel_size)_. See under the **variables** header on Pytorch's page on Conv2d [here](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html). I think this clarifies the confusion.

Best,
Shibhansh

**Comment by kiranchari** [comment]
Thanks for both clarifications @shibhansh.
